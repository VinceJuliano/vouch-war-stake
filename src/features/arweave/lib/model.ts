import { z } from "zod";

export const ArweaveId = z.string().regex(/^[a-zA-Z0-9_-]{43}$/);
export type ArweaveId = z.infer<typeof ArweaveId>;

export const ArweaveAddress = ArweaveId;
export type ArweaveAddress = ArweaveId;

export const ArweaveTxId = ArweaveId;
export type ArweaveTxId = ArweaveId;
