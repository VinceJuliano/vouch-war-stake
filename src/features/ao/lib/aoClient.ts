import { connect, result } from "@permaweb/aoconnect";

export type MessageId = string;

type BaseAoClient = ReturnType<typeof connect>;

export type AoClient = BaseAoClient;

export type MessageResult = Awaited<ReturnType<typeof result>>;

// export type Message = MessageResult["Messages"][0];
export type Message = {
  Target: string;
  Tags: Array<{
    name: string;
    value: string;
  }>;
  Data: string;
};

export type MessageWithResult = {
  messageId: MessageId;
  result: MessageResult;
};
