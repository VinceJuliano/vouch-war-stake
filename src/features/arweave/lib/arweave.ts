import { base32, base64urlnopad } from "@scure/base";
import Arweave from "arweave";

export const defaultArweave = Arweave.init({});

export function fetchUrl(txId: string) {
  const txIdData = base64urlnopad.decode(txId);
  const base32UnpaddedTxId = base32.encode(txIdData).replace(/=/g, "");
  return `https://${base32UnpaddedTxId}.arweave.net/${txId}`;
}
