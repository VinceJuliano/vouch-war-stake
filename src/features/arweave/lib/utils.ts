export const truncateAddress = (
  address: string,
  prefixLength: number = 6,
  suffixLength: number = 4,
  middleText: string = "...",
) => {
  return `${address.slice(0, prefixLength)}${middleText}${address.slice(-suffixLength)}`;
};
