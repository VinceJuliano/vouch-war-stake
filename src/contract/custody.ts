import { dryrun, message } from "@permaweb/aoconnect";
import { MutationOptions, queryOptions } from "@tanstack/react-query";
import { getTagValue } from "../lib/arweave";
import { AoSigner } from "../hooks/useAoSigner";

export type DepositParameters = {
  tokenId: string;
  quantity: string;
  stakeDurationMs: number;
};

export const getInfoQuery = (processId: string) =>
  queryOptions({
    queryKey: ["Custody", processId, "Info"],
    queryFn: async () => {
      const res = await dryrun({
        process: processId,
        tags: [
          {
            name: "Action",
            value: "Info",
          },
        ],
      });
      const error = getTagValue(res.Messages[0].Tags, "Error");
      if (error) {
        return error;
      }
      return JSON.parse(res.Messages[0].Data);
    },
  });

export const getActiveStakesQuery = (processId: string, walletId?: string) =>
  queryOptions({
    queryKey: ["Custody", processId, "Get-Active-Stakes"],
    queryFn: async () => {
      const res = await dryrun({
        process: processId,
        tags: [
          {
            name: "Action",
            value: "Custody.Get-Active-Stakes",
          },
          ...(walletId
            ? [
                {
                  name: "Wallet-Id",
                  value: walletId,
                },
              ]
            : []),
        ],
      });
      const error = getTagValue(res.Messages[0].Tags, "Error");
      if (error) {
        return error;
      }
      return JSON.parse(res.Messages[0].Data);
    },
  });

export const depositMutation = (
  custodyProcessId: string,
  { tokenId, quantity, stakeDurationMs }: DepositParameters,
  aoSigner: AoSigner,
): MutationOptions => ({
  mutationKey: ["Custody", custodyProcessId, "Deposit", tokenId],
  mutationFn: async () => {
    const messageId = await message({
      process: tokenId,
      tags: [
        {
          name: "Action",
          value: "Transfer",
        },
        {
          name: "Recipient",
          value: custodyProcessId,
        },
        {
          name: "Quantity",
          value: quantity,
        },
        {
          name: "X-Stake-Duration",
          value: stakeDurationMs.toString(),
        },
      ],
      signer: aoSigner,
    });
    console.log({ messageId });
    return messageId;
  },
});
