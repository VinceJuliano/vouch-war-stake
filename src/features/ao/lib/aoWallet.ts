import { createDataItemSigner } from "@permaweb/aoconnect";
import { WalletType } from "./wallets";

export type AoSigner = ReturnType<typeof createDataItemSigner>;

export type AoWallet = {
  type: WalletType;
  anonymous: boolean;
  address: string;
  signer: AoSigner;
};

export type AoWalletConnectionResult =
  | {
      success: true;
      result: AoWallet;
      disconnect?: () => void;
    }
  | {
      success: false;
      error: unknown;
    };

// TODO: Import these from somewhere more official
export type PermissionType =
  | "ACCESS_ADDRESS"
  | "ACCESS_PUBLIC_KEY"
  | "ACCESS_ALL_ADDRESSES"
  | "SIGN_TRANSACTION"
  | "ENCRYPT"
  | "DECRYPT"
  | "SIGNATURE"
  | "ACCESS_ARWEAVE_CONFIG"
  | "DISPATCH";

export interface AppInfo {
  name?: string;
  logo?: string;
}

export interface GatewayConfig {
  host: string;
  port: number;
  protocol: "http" | "https";
}

export type ConnectConfig = {
  permissionsRequested: PermissionType[];
  permissionsRequired?: PermissionType[];
  appInfo?: AppInfo;
  gateway?: GatewayConfig;
};

export type AoWalletConnector = (
  config: ConnectConfig,
  onDisconnect?: () => void,
) => Promise<AoWalletConnectionResult>;
