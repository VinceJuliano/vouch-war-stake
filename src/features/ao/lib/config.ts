import { PermissionType } from "./aoWallet";

export const permissionsRequired: Array<PermissionType> = [
  "ACCESS_ADDRESS",
  "SIGN_TRANSACTION",
  // "ACCESS_PUBLIC_KEY",
  // "ENCRYPT",
  // "DECRYPT",
  "SIGNATURE",
  "ACCESS_ARWEAVE_CONFIG",
];

export const permissionsRequested = [
  ...permissionsRequired,
  // ...optionalPermissions,
];
