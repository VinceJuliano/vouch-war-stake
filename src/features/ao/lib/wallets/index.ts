import { AoWalletConnector } from "../aoWallet";
import { connectInjectedWallet } from "./injected";
import { connectOthentWallet } from "./othent";
import { createGeneratedWallet } from "./generated";

export const WalletType = ["Injected", "Othent", "Keyfile"] as const;
export type WalletType = (typeof WalletType)[number];

export const wallets: Record<WalletType, AoWalletConnector> = {
  Injected: connectInjectedWallet,
  Othent: connectOthentWallet,
  Keyfile: createGeneratedWallet,
} as const;
