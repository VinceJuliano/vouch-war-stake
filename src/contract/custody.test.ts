import { describe, it, expect } from "vitest";
import { getInfoQuery, getActiveStakesQuery } from "./custody";
import { queryClient } from "../lib/query";

const testContractId = "48Bg_s9zmZzLKlvpeXPs15qID2uL1xNzJ89sPz5VPhM";

describe("Custody contract", () => {
  it("getInfoQuery", async () => {
    const res = await queryClient.fetchQuery(getInfoQuery(testContractId));
    expect(res).toMatchSnapshot();
  });

  it("getActiveStakesQuery", async () => {
    const res = await queryClient.fetchQuery(
      getActiveStakesQuery(testContractId),
    );
    expect(res).toMatchSnapshot();
  });
});
