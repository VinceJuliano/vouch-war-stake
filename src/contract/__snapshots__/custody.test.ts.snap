// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Custody contract > getActiveStakesQuery 1`] = `
{
  "ActiveStakes": [
    {
      "DepositTimeMs": 1726383444675,
      "Id": "azpc7I6H5R4nIWY5nsNeK_WUV_pGlilQ1hSSdqV9Zyw",
      "Quantity": "1000000000",
      "Sender": "0cQJ5Hd4oCaL57CWP-Pqe5_e0D4_ZDWuxKBgR9ke1SI",
      "StakeDurationMs": 86400000,
      "TokenId": "xU9zFkq3X2ZQ6olwNVvr1vUWIjc3kXTWr7xKQD6dh10",
      "WithdrawTimeMs": 1726469844675,
      "timestamp": 1726383444675,
    },
  ],
}
`;

exports[`Custody contract > getInfoQuery 1`] = `
{
  "RewardTokens": {
    "m3PaWzK4PTG9lAaqYQPaPdOcXdO8hYqi5Fe9NWqXd0w": {
      "Enabled": true,
    },
  },
  "StakingTokens": {
    "xU9zFkq3X2ZQ6olwNVvr1vUWIjc3kXTWr7xKQD6dh10": {
      "DurationBounds": {
        "Max": 31536000000,
        "Min": 60000,
      },
      "Enabled": true,
      "QuantityBounds": {
        "Max": "1000000000000000",
        "Min": "1000000000",
      },
    },
  },
}
`;
